### DISCUSSION TITLE
W3C compliant locators to use in Appium Espresso Driver with Selenium Grid 4

### DISCUSSION THREAD

**Discussion by ValzRon** [appium/java-client#1798] [Q&A]
Need some help here. Trying to migrate my Appium 2.0 setup to use Selenium Grid 4. Problem is, due to W3C compliance it doesn't support certain locators anymore like By.Id . Ive opened an issue on Selenium Repo and got a response:

https://github.com/SeleniumHQ/seleniumhq.github.io/issues/1224#issue-1444314996

**My question is:** What is the best practice for locating elements in an Android (Espresso) context? Im using AndroidFindBy ID in hundreds of instances, so changing all to XPath would be hellish. Thnx in advance!

**Comment by valfirst** [comment]
Use `io.appium.java_client.AppiumBy#id` instead of `org.openqa.selenium.By#id`.

> Im using AndroidFindBy ID

And make sure you use Appium Java Client of version 8.0.0 or higher.

**Reply by ValzRon** [reply]
So Im using Appium Java client 8.2.0 and I tried to turn this

```
@AndroidFindBy(id = "imageView_icon_dashboardTile")
private List<WebElement> dashboardTitleIcons;
```

into this (getSession calls driver instance)

`private List<WebElement> dashboardTitleIcons = getSession().findElements(AppiumBy.id("imageView_icon_dashboardTile"));`

but I get this error, seems to be css selector still:

Error creating bean with name 'scopedTarget.dashboardActivityPage' defined in file [C:\Users\MyUser\IdeaProjects\appium-tests\build\classes\java\test\page\DashboardActivityPage.class]: Instantiation of bean failed; nested exception is org.springframework.beans.BeanInstantiationException: Failed to instantiate [page.DashboardActivityPage]: Constructor threw exception; nested exception is org.openqa.selenium.ScriptTimeoutException: Unable to execute request for an existing session: Strategy is not implemented: css selector

**Reply by valfirst** [reply]
Could you please provide [Short, Self Contained, Correct (Compilable), Example](http://sscce.org/)?

**Reply by ValzRon** [reply]
Here im trying to create a List of WebElements to represent multiple icons of the same id. Then I retrieve the first one and click it. The click happens inside a Cucumber step class.
```
@Component
@ScenarioScope
public class DashboardActivityPage extends BaseActivityPage {
....
    private List<WebElement> dashboardTitleIcons = getSession().findElements(AppiumBy.id("imageView_icon_dashboardTile"));
...
    private WebElement getDashboardIcon() {
        return dashboardTitleIcons.get(0);
    }
...
    public void clickDashboardIcon() {
        getDashboardIcon().click();
    }
}
```
`getSession `gets me the instance of driver (AndroidDriver) from the `BaseActivityPage `. For that im using the newest Espresso driver 2.13.5 and EspressoOptions to fill caps. Also Im using Appium 2.0.0-beta.46 and Selenium 4.5.0 with Grid4.

All of this actually works fine if im using the Appium Server directly, but in conjunction with Grid 4 it doesnt work. Ive configured a Relay to create a communication between Appium Server and Grid 4 and it does receive everything fine.

Then I get the above mentioned error, here are some additional logs:
```
For documentation on this error, please visit: https://selenium.dev/exceptions/#invalid_selector_exception
Build info: version: '4.5.0', revision: 'fe167b119a'
System info: os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '1.8.0_351'
Driver info: org.openqa.selenium.grid.node.CustomLocatorHandler$CustomWebDriver
Command: [eb3e6567-6310-44ec-86b9-b3b091c70cfa, findElements {using=id, value=imageView_icon_dashboardTile}]
Capabilities {}
Session ID: eb3e6567-6310-44ec-86b9-b3b091c70cfa
Build info: version: '4.5.0', revision: 'fe167b119a'
System info: os.name: 'Windows 10', os.arch: 'amd64', os.version: '10.0', java.version: '11.0.17'
Driver info: io.appium.java_client.android.AndroidDriver
Command: [eb3e6567-6310-44ec-86b9-b3b091c70cfa, findElements {using=id, value=imageView_icon_dashboardTile}]
```

**Reply by valfirst** [reply]
Unfortunately this description doesn't look like something that can be easily run by anyone to reproduce the issue, please, read the provided guideline carefully in order to assemble short and self-contained project to run.
